### PR TITLE
Unit test update for multiple java compilers [TG-2483]

### PR DESCRIPTION
**Pull request by thomasspriggs**
This is a modification to an existing unit test, to re-run the tests using class files which have been generated by various different java compilers.

### CODE REVIEW

**Comment by thomasspriggs** [comment]
@thk123 Review this please.

**Review by thk123 — APPROVED** [review]

**Review by tautschnig — CHANGES_REQUESTED** [review]
I'm sure the changes in this PR are good and useful, but some further information is necessary for legal and maintenance purposes: how have all these binary files been generated and, which might then already be answered, is it actually legal to distribute them in this form? The answer should most likely be in the form of an extended commit message and/or a README file. Thanks!

**Comment by thomasspriggs** [comment]
@tautschnig I have added a readme file, about generating the class files. The class files are the output from compiling our java files with various java compilers. They are not taken from the internals of any java compiler or internals of any java run time environment / library.

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
> I have added a readme file [...]

Brilliant, thank you!

File: unit/java_bytecode/java_bytecode_parser/lambda_examples/readme.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ 3. Use `javac -version` to confirm the desired compilier has been correctly configured. The version information for each compiler is as follows -
+    * openJdk 8     - 1.8.0_151
+    * oracle java 8 - 1.8.0_161
+    * oracle java 9 - 9.0.4
+ 4. Compile the `.java` files using `javac -g ./*.java`. The `-g` flag switches on debugging output, which we require.
+ 5. The `.class` files should have been written to the current directory. They can now be moved into the appropriate sub directory. For example - `mkdir my_compiler_classes && mv ./*.class ./my_compiler_classes`
+ 6. If this is a new compiler, add it to the test, by adding the compilier to the list in the `run_test_with_compilers` function in `java_bytecode_parse_lambda_method_table.cpp`.
+ 7. Repeat steps 2 through 6 for each additional compiler.
+
+The proceedure for compiling with eclipse is different. Installing the java eclipse ide does not setup a version of javac command. So to produce the class files for eclipse, the proceedure is as follows -
```
Comment (thk123) [inline unit/java_bytecode/java_bytecode_parser/lambda_examples/readme.md]: Nit: eclipse should be capitalised, IDE should be all caps

File: unit/java_bytecode/java_bytecode_parser/lambda_examples/readme.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+Each sub directory inside this one contains class files which have been compiled with specific java compilers. The proceedure taken to produce these files is as follows -
+ 1. Install the java compilers of interest. Skip this step if you already have them installed.
+ 2. Use `sudo update-alternatives --config javac` to configure your system to use the compiler of interest.
+ 3. Use `javac -version` to confirm the desired compilier has been correctly configured. The version information for each compiler is as follows -
+    * openJdk 8     - 1.8.0_151
+    * oracle java 8 - 1.8.0_161
+    * oracle java 9 - 9.0.4
+ 4. Compile the `.java` files using `javac -g ./*.java`. The `-g` flag switches on debugging output, which we require.
```
Comment (thk123) [inline unit/java_bytecode/java_bytecode_parser/lambda_examples/readme.md]: You can use `-d` on javac to specify the output directory, but it requires it to exist first, so don't know if it actually makes it much easier to follow the instructions...

File: unit/java_bytecode/java_bytecode_parser/lambda_examples/readme.md (hunk 1 above)
Comment (thk123) [inline unit/java_bytecode/java_bytecode_parser/lambda_examples/readme.md]: As above you should probably mention what version of Eclipse you are using.

**Review by thk123 — APPROVED** [review]

File: unit/java_bytecode/java_bytecode_parser/lambda_examples/readme.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+Each sub directory inside this one contains class files which have been compiled with specific java compilers. The proceedure taken to produce these files is as follows -
```
Comment (peterschrammel) [inline unit/java_bytecode/java_bytecode_parser/lambda_examples/readme.md]: procedure

File: unit/java_bytecode/java_bytecode_parser/lambda_examples/readme.md (hunk 1 above)
Comment (peterschrammel) [inline unit/java_bytecode/java_bytecode_parser/lambda_examples/readme.md]: procedure

Comment (peterschrammel) [reply unit/java_bytecode/java_bytecode_parser/lambda_examples/readme.md]: You should be able to run the Eclipse compiler as explained here: https://help.eclipse.org/neon/index.jsp?topic=%2Forg.eclipse.jdt.doc.user%2Ftasks%2Ftask-using_batch_compiler.htm

**Comment by hannes-steffenhagen-diffblue** [comment]
Instead of a lengthy description, wouldn't it be easier (going forward) to include a Maven or Gradle build script that generates these class files? Both of them respect JAVA_HOME, so you don't need update-alternatives to pick different JDKs for those.

**Comment by thomasspriggs** [comment]
@hannes-steffenhagen-diffblue I will bear in mind the option of using some type of build script for future instances of this type of use case. I would probably use a bash script instead of a maven file as there are no dependencies to resolve. Can JAVA_HOME be set from inside a Maven/Gradle build script or would a separate bash script be required to set this and run for each compiler anyway?

**Pull request merged**